File: zima-file-service/Api/ResponseCache.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace ZimaFileService.Api;

/// <summary>
/// Response cache for storing and retrieving cached prompt responses.
/// Reduces response time for repeated or similar prompts.
/// </summary>
public class ResponseCache
{
    private static readonly Lazy<ResponseCache> _instance = new(() => new ResponseCache());
    public static ResponseCache Instance => _instance.Value;

    private readonly ConcurrentDictionary<string, CacheEntry> _cache = new();
    private readonly TimeSpan _defaultTtl = TimeSpan.FromMinutes(30);
    private readonly int _maxCacheSize = 1000;
    private readonly Timer _cleanupTimer;

    // Cache statistics
    public int CacheHits { get; private set; }
    public int CacheMisses { get; private set; }
    public double HitRate => CacheHits + CacheMisses > 0
        ? (double)CacheHits / (CacheHits + CacheMisses) * 100
        : 0;

    private ResponseCache()
    {
        // Cleanup expired entries every 5 minutes
        _cleanupTimer = new Timer(CleanupExpiredEntries, null, TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(5));
        Log("INFO", "ResponseCache initialized");
    }

    /// <summary>
    /// Generate a cache key from prompt and context.
    /// Uses SHA256 hash for consistent key generation.
    /// </summary>
    public static string GenerateCacheKey(string prompt, string? sessionId = null, int messageCount = 0)
    {
        // Normalize prompt (trim, lowercase for comparison)
        var normalizedPrompt = prompt.Trim().ToLowerInvariant();

        // Include context that affects the response
        var keySource = $"{normalizedPrompt}|{messageCount}";

        using var sha256 = SHA256.Create();
        var hashBytes = sha256.ComputeHash(Encoding.UTF8.GetBytes(keySource));
        return Convert.ToHexString(hashBytes)[..16]; // Use first 16 chars
    }

    /// <summary>
    /// Try to get a cached response.
    /// </summary>
    public bool TryGet(string cacheKey, out ZimaResponse? response)
    {
        if (_cache.TryGetValue(cacheKey, out var entry))
        {
            if (entry.ExpiresAt > DateTime.UtcNow)
            {
                CacheHits++;
                response = entry.Response;
                Log("DEBUG", $"Cache HIT: {cacheKey} (hits: {CacheHits}, rate: {HitRate:F1}%)");
                return true;
            }

            // Entry expired, remove it
            _cache.TryRemove(cacheKey, out _);
        }

        CacheMisses++;
        response = null;
        Log("DEBUG", $"Cache MISS: {cacheKey} (misses: {CacheMisses}, rate: {HitRate:F1}%)");
        return false;
    }

    /// <summary>
    /// Store a response in the cache.
    /// </summary>
    public void Set(string cacheKey, ZimaResponse response, TimeSpan? ttl = null)
    {
        // Don't cache failed responses or responses with files (files may change)
        if (!response.Success || response.GeneratedFiles.Count > 0)
        {
            Log("DEBUG", $"Cache SKIP: {cacheKey} (success={response.Success}, files={response.GeneratedFiles.Count})");
            return;
        }

        // Enforce max cache size
        if (_cache.Count >= _maxCacheSize)
        {
            EvictOldestEntries(_maxCacheSize / 10); // Evict 10%
        }

        var entry = new CacheEntry
        {
            Response = response,
            CreatedAt = DateTime.UtcNow,
            ExpiresAt = DateTime.UtcNow.Add(ttl ?? _defaultTtl),
            CacheKey = cacheKey
        };

        _cache[cacheKey] = entry;
        Log("DEBUG", $"Cache SET: {cacheKey} (size: {_cache.Count}, ttl: {(ttl ?? _defaultTtl).TotalMinutes}min)");
    }

    /// <summary>
    /// Check if a prompt is cacheable (simple queries are more cacheable).
    /// </summary>
    public static bool IsCacheable(string prompt)
    {
        var lowerPrompt = prompt.ToLower();

        // Don't cache file generation requests
        if (lowerPrompt.Contains("create") || lowerPrompt.Contains("generate") ||
            lowerPrompt.Contains("make") || lowerPrompt.Contains("build"))
            return false;

        // Don't cache modification requests
        if (lowerPrompt.Contains("update") || lowerPrompt.Contains("modify") ||
            lowerPrompt.Contains("change") || lowerPrompt.Contains("edit"))
            return false;

        // Cache simple questions and queries
        if (lowerPrompt.StartsWith("what") || lowerPrompt.StartsWith("how") ||
            lowerPrompt.StartsWith("why") || lowerPrompt.StartsWith("explain") ||
            lowerPrompt.StartsWith("list") || lowerPrompt.StartsWith("describe"))
            return true;

        // Cache short prompts (likely simple queries)
        if (prompt.Length < 100)
            return true;

        return false;
    }

    /// <summary>
    /// Get cache statistics.
    /// </summary>
    public CacheStats GetStats()
    {
        return new CacheStats
        {
            TotalEntries = _cache.Count,
            CacheHits = CacheHits,
            CacheMisses = CacheMisses,
            HitRate = HitRate,
            MaxSize = _maxCacheSize
        };
    }

    /// <summary>
    /// Clear the entire cache.
    /// </summary>
    public void Clear()
    {
        _cache.Clear();
        CacheHits = 0;
        CacheMisses = 0;
        Log("INFO", "Cache cleared");
    }

    private void CleanupExpiredEntries(object? state)
    {
        var now = DateTime.UtcNow;
        var expiredKeys = _cache
            .Where(kvp => kvp.Value.ExpiresAt <= now)
            .Select(kvp => kvp.Key)
            .ToList();

        foreach (var key in expiredKeys)
        {
            _cache.TryRemove(key, out _);
        }

        if (expiredKeys.Count > 0)
        {
            Log("DEBUG", $"Cache cleanup: removed {expiredKeys.Count} expired entries");
        }
    }

    private void EvictOldestEntries(int count)
    {
        var oldestKeys = _cache
            .OrderBy(kvp => kvp.Value.CreatedAt)
            .Take(count)
            .Select(kvp => kvp.Key)
            .ToList();

        foreach (var key in oldestKeys)
        {
            _cache.TryRemove(key, out _);
        }

        Log("DEBUG", $"Cache eviction: removed {oldestKeys.Count} oldest entries");
    }

    private static void Log(string level, string message)
    {
        Console.WriteLine($"[{DateTime.Now:HH:mm:ss.fff}] [CACHE] [{level}] {message}");
    }
}

public class CacheEntry
{
    public ZimaResponse Response { get; set; } = new();
    public DateTime CreatedAt { get; set; }
    public DateTime ExpiresAt { get; set; }
    public string CacheKey { get; set; } = "";
}

public class CacheStats
{
    public int TotalEntries { get; set; }
    public int CacheHits { get; set; }
    public int CacheMisses { get; set; }
    public double HitRate { get; set; }
    public int MaxSize { get; set; }
}

File: gateway/src/context/response-cache.ts
/**
 * Response Cache Manager
 * Caches responses to avoid redundant API calls (ZIMA approach)
 */

import { createHash } from 'crypto';
import { MessageResponse, CachedResponse } from '../types';

export class ResponseCache {
  private cache: Map<string, CachedResponse>;
  private ttl: number;
  private maxSize: number;
  private hits: number;
  private misses: number;

  constructor(ttl: number = 3600000, maxSize: number = 1000) {
    this.cache = new Map();
    this.ttl = ttl; // Default: 1 hour
    this.maxSize = maxSize;
    this.hits = 0;
    this.misses = 0;

    // Clean up expired entries every 5 minutes
    setInterval(() => this.cleanup(), 300000);
  }

  /**
   * Generate cache key from request parameters
   */
  generateCacheKey(
    message: string,
    sessionKey: string,
    messageCount: number,
    attachmentCount: number = 0
  ): string {
    // Create hash from key components
    const keyData = `${message}:${sessionKey}:${messageCount}:${attachmentCount}`;
    return createHash('sha256').update(keyData).digest('hex');
  }

  /**
   * Get cached response
   */
  get(cacheKey: string): MessageResponse | null {
    const cached = this.cache.get(cacheKey);

    if (!cached) {
      this.misses++;
      return null;
    }

    // Check if expired
    const age = Date.now() - cached.timestamp;
    if (age > this.ttl) {
      this.cache.delete(cacheKey);
      this.misses++;
      return null;
    }

    this.hits++;
    console.log(`💾 Cache HIT (${this.getHitRate().toFixed(1)}% hit rate)`);

    return {
      ...cached.response,
      fromCache: true
    };
  }

  /**
   * Set cached response
   */
  set(cacheKey: string, response: MessageResponse): void {
    // Check cache size limit
    if (this.cache.size >= this.maxSize) {
      this.evictOldest();
    }

    this.cache.set(cacheKey, {
      response,
      timestamp: Date.now(),
      cacheKey
    });

    console.log(`💾 Cached response (cache size: ${this.cache.size}/${this.maxSize})`);
  }

  /**
   * Check if key exists in cache
   */
  has(cacheKey: string): boolean {
    const cached = this.cache.get(cacheKey);
    if (!cached) return false;

    const age = Date.now() - cached.timestamp;
    return age <= this.ttl;
  }

  /**
   * Clear cache
   */
  clear(): void {
    this.cache.clear();
    this.hits = 0;
    this.misses = 0;
    console.log('💾 Cache cleared');
  }

  /**
   * Get cache statistics
   */
  getStats(): {
    size: number;
    maxSize: number;
    hits: number;
    misses: number;
    hitRate: number;
    oldestEntry: number;
  } {
    let oldestTimestamp = Date.now();

    for (const cached of this.cache.values()) {
      if (cached.timestamp < oldestTimestamp) {
        oldestTimestamp = cached.timestamp;
      }
    }

    return {
      size: this.cache.size,
      maxSize: this.maxSize,
      hits: this.hits,
      misses: this.misses,
      hitRate: this.getHitRate(),
      oldestEntry: Date.now() - oldestTimestamp
    };
  }

  /**
   * Get cache hit rate percentage
   */
  private getHitRate(): number {
    const total = this.hits + this.misses;
    return total === 0 ? 0 : (this.hits / total) * 100;
  }

  /**
   * Clean up expired entries
   */
  private cleanup(): void {
    const now = Date.now();
    let removed = 0;

    for (const [key, cached] of this.cache.entries()) {
      const age = now - cached.timestamp;
      if (age > this.ttl) {
        this.cache.delete(key);
        removed++;
      }
    }

    if (removed > 0) {
      console.log(`💾 Cache cleanup: removed ${removed} expired entries`);
    }
  }

  /**
   * Evict oldest entry when cache is full
   */
  private evictOldest(): void {
    let oldestKey: string | null = null;
    let oldestTimestamp = Date.now();

    for (const [key, cached] of this.cache.entries()) {
      if (cached.timestamp < oldestTimestamp) {
        oldestTimestamp = cached.timestamp;
        oldestKey = key;
      }
    }

    if (oldestKey) {
      this.cache.delete(oldestKey);
      console.log('💾 Evicted oldest cache entry');
    }
  }
}

// Singleton instance
let responseCache: ResponseCache | null = null;

/**
 * Get global response cache instance
 */
export function getResponseCache(): ResponseCache {
  if (!responseCache) {
    responseCache = new ResponseCache();
  }
  return responseCache;
}
